### ISSUE TITLE
Subpixel artifacts when drawing on canvas

### ISSUE THREAD

**Issue by EzzPress** [JetBrains/compose-multiplatform#3068]
**Bug**
I'm drawing something like heatmap chart (x,y and z as color). when number of elements grows. I got something like subpixel artifacts as edges between drawn rectangles, the brightness and position of the artifact lines depends of window size.
I was trying to workaround with blur, colorfilters, draw with skia funtions, enabling antialias in Paint. Nothing really helped. 

**Affected platforms**
- Desktop

**Versions**
- Kotlin version*: 1.8.0
- Compose Multiplatform version*:  1.3.0
- OS version(s)* (required for Desktop and iOS issues):  MacOs, Linux, Windows
- OS architecture (x86 or arm64):  Apple Silicon M1, x86
- JDK (for desktop issues): 11

**To Reproduce**
Steps and/or the code snippet to reproduce the behavior:
1. Create composable with canvas with white background
2. Draw black rectangles side-by-side to fill all the canvas, size must be fractional (for example - 50.5f)
3. Run app with created composable
4. See the artifacts
5. Change window side by horizontal direction - state of artifacts changes
6. Change window side by vertical direction - state of artifacts changes

**Expected behavior**
Solid color of canvas filled by rectangles

**Screenshots**
![image](https://user-images.githubusercontent.com/109756293/232457738-8bcdb7f4-505c-4b47-a440-485c4c846e49.png)

**Comment by dima-avdeev-jb** [comment]
Can you please provide a code snippet or reproducible sample on GitHub?

**Comment by EzzPress** [comment]
https://github.com/EzzPress/heatmapcompose here it is

**Comment by dima-avdeev-jb** [comment]
Made workaround here:
https://github.com/EzzPress/heatmapcompose/pull/1

**Comment by dima-avdeev-jb** [comment]
Android has the same behaviour. I think it is not a bug.

Minimal reproducer:
```Kotlin
Canvas(modifier = Modifier.fillMaxSize().background(Color.White)) {
        val chartWidth = size.width
        val chartHeight = size.height
        var x = 0f
//        val rectSize: Float = 50f // works fine
        val rectSize: Float = 50.5f // Show grid
        while (x < chartWidth) {
            var y = 0f
            while (y < chartHeight) {
                drawRect(Color.Black, Offset(x, y), Size(rectSize, rectSize))
                y += rectSize
            }
            x += rectSize
        }
    }
```

**Comment by dima-avdeev-jb** [comment]
Can you apply this workaround?
https://github.com/EzzPress/heatmapcompose/pull/1

**Comment by dima-avdeev-jb** [comment]
Ok, I will reopen this issue. But I will change the description.

**Comment by dima-avdeev-jb** [comment]
@EzzPress Can you please check, that changed description is correct?

**Comment by EzzPress** [comment]
No, this workaround doesn't solve the problem, because it doesn't solve the float mapping,
obviously the skia somewhere inside makes the same operations of rounding.

![image](https://user-images.githubusercontent.com/109756293/235143140-6052f2f5-acce-4175-8e65-309fc0e07ce2.png)

To get a quality image we need some how to transform initial m x n matrix  to visual matrix k x l with data reducing.
In javafx in the jfreechart lib, it plots well, but the time performance is very low, and all the dots are present and could generate onActions, that also ruins performance.

**Comment by dima-avdeev-jb** [comment]
Can you please check that the changed description of this issue is correct?

**Comment by EzzPress** [comment]
> Can you please check that the changed description of this issue is correct?

I'd say - Absence of smoothing algorithm for subpixel artifacts when drawing on Canvas with fractional value coordinates

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-3068">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
